chunk/aws_storage_client.go

```diff
@@ -61,41 +66,83 @@ var (
 		Name:      "dynamo_failures_total",
 		Help:      "The total number of errors while storing chunks to the chunk store.",
 	}, []string{tableNameLabel, errorReasonLabel})
+	s3RequestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "s3_request_duration_seconds",
+		Help:      "Time spent doing S3 requests.",
+		Buckets:   []float64{.025, .05, .1, .25, .5, 1, 2},
+	}, []string{"operation", "status_code"})
 )
 
 func init() {
 	prometheus.MustRegister(dynamoRequestDuration)
 	prometheus.MustRegister(dynamoConsumedCapacity)
 	prometheus.MustRegister(dynamoFailures)
+	prometheus.MustRegister(s3RequestDuration)
 }
 
-type dynamoClientAdapter struct {
-	DynamoDB dynamodbiface.DynamoDBAPI
+// AWSStorageConfig specifies config for storing data on AWS.
+type AWSStorageConfig struct {
+	DynamoDB util.URLValue
+	S3       util.URLValue
 }
 
-// NewDynamoDBClient makes a new DynamoDBClient
-func NewDynamoDBClient(dynamoDBURL string) (StorageClient, string, error) {
-	url, err := url.Parse(dynamoDBURL)
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *AWSStorageConfig) RegisterFlags(f *flag.FlagSet) {
+	f.Var(&cfg.DynamoDB, "dynamodb.url", "DynamoDB endpoint URL with escaped Key and Secret encoded. "+
+		"If only region is specified as a host, proper endpoint will be deduced. Use inmemory:///<table-name> to use a mock in-memory implementation.")
+	f.Var(&cfg.S3, "s3.url", "S3 endpoint URL with escaped Key and Secret encoded. "+
+		"If only region is specified as a host, proper endpoint will be deduced. Use inmemory:///<bucket-name> to use a mock in-memory implementation.")
+}
+
+type awsStorageClient struct {
+	DynamoDB   dynamodbiface.DynamoDBAPI
+	S3         *s3.S3
+	bucketName *string
+}
+
+// NewAWSStorageClient makes a new AWS-backed StorageClient.
+func NewAWSStorageClient(cfg AWSStorageConfig) (StorageClient, string, error) {
+	tableName := strings.TrimPrefix(cfg.DynamoDB.URL.Path, "/")
+
+	if cfg.DynamoDB.URL.Scheme == "inmemory" {
+		return NewMockStorage(), tableName, nil
+	}
+
+	dynamoDBConfig, err := awsConfigFromURL(cfg.DynamoDB.URL)
 	if err != nil {
 		return nil, "", err
 	}
 
-	dynamoDBConfig, err := awsConfigFromURL(url)
+	dynamoDB := dynamodb.New(session.New(dynamoDBConfig))
+	if cfg.S3.URL == nil {
+		storageClient := awsStorageClient{
+			DynamoDB: dynamoDB,
+		}
+		return storageClient, tableName, nil
+	}
+
+	s3Config, err := awsConfigFromURL(cfg.S3.URL)
 	if err != nil {
 		return nil, "", err
 	}
+	s3Client := s3.New(session.New(s3Config))
+	bucketName := aws.String(strings.TrimPrefix(cfg.S3.URL.Path, "/"))
 
-	dynamoDBClient := dynamoClientAdapter{dynamodb.New(session.New(dynamoDBConfig))}
-	tableName := strings.TrimPrefix(url.Path, "/")
-	return dynamoDBClient, tableName, nil
+	storageClient := awsStorageClient{
+		DynamoDB:   dynamoDB,
+		S3:         s3Client,
+		bucketName: bucketName,
+	}
+	return storageClient, tableName, nil
```

Comment:
It doesn't need to be in the storage client at all. Have added a flag to the SchemaConfig, removed the return value from `NewAWSStorageClient`, and deprecated the parameter to `NewChunkStore`.